engine/src/tangl/core/graph/graph.py

```
@@ -78,9 +82,34 @@ def ancestors(self) -> Iterator[Subgraph]:
+
+        ancestors = [self] + list(self.ancestors())
+        ancestor_tags = { t for a in ancestors for t in a.tags }
+        match_logger.debug(f"Comparing query tags {tags} against {ancestor_tags}")
+        return tags.issubset(ancestor_tags)
+
+    def has_parent_label(self, parent_label: str) -> bool:
+        # seems redundant
+        return self.parent is not None and self.parent.label == parent_label
+
+    def has_scope(self, scope: dict) -> bool:
+        warnings.warn(DeprecationWarning("`has_scope` is deprecated; prefer `matches(has_path, has_ancestor_tags`)."))
```

Ruff (0.14.10)

No explicit `stacklevel` keyword argument found

Set `stacklevel=2`

(B028)